icu4c/source/tools/toolutil/writesrc.cpp

```diff
@@ -395,3 +461,29 @@ usrc_writeArrayOfMostlyInvChars(FILE *f,
         fputs(postfix, f);
     }
 }
+
+U_CAPI void U_EXPORT2
+usrc_writeString(FILE *f,
+                 const UChar* ptr, int32_t length,
+                 UTargetSyntax) {
+    UnicodeString ustr(ptr, length);
+    // For now, assume all UTargetSyntax values are valid here.
+    fprintf(f, "\"");
+    for (int32_t i = 0; i < ustr.length(); ) {
+        UChar32 cp = ustr.char32At(i);
+        if (cp == u'"') {
+            fprintf(f, "\\\"");
+        } else if (ICU_Utility::isUnprintable(cp)) {
+            UnicodeString u16result;
+            ICU_Utility::escapeUnprintable(u16result, cp);
+            std::string u8result;
+            u16result.toUTF8String(u8result);
+            fprintf(f, "%s", u8result.data());
+        } else {
+            char s[2] = {static_cast<char>(cp), 0};
+            fprintf(f, "%s", s);
+        }
+        i += U8_LENGTH(cp);
```

Comment:
This can only work with ASCII :-)

```suggestion
        i += U16_LENGTH(cp);
```
... unnecessary if you use U16_NEXT() instead.